Improved NCL interface

bouweandela opened the pull request:
Note that this PR is against branch `version2_multiple_var_info`, not against the `version2_development` branch.

@mattiarighi This is how I would propose the make the NCL interface. It is much more robust than the previous implementation.

Improvements include
- Use the NCL List type to represent Python/yaml lists instead of arrays. This is much more flexible, since the NCL List type supports items of different type.
- One item in the `input_file_info` list represents one variable of one dataset. This way much less use of indexes is required, because all information belonging to one preprocessed file is stored in the attributes of one NCL logical. Less index usage means more readable code and thus lower chances of bugs.
- None values are now more correctly handled (stored as _Missing).
- This implementation has a lower risk of accidentally writing invalid NCL.
- The input_file_info list is the same as the information that Python and R scripts receive, so it will be easier for diagnostic developers to switch from one language to another: no need to re-learn the interface.

mattiarighi commented:
There is a problem: this is now creating an `input_file_info` list element for each dataset and each variable, which screws up the indexing in all NCL scripts.

Was it meant to work this way?
This would mean a recoding and testing of basically all NCL scripts.

bouweandela commented:
Yes, it is meant to work that way. Because if you implement it like this, you can get rid of the indices almost everywhere, which makes the code more readable and less error prone (the second point in the bullet list above).

However, you are right that it requires getting rid of all the indices almost everywhere so it requires some work on your part, which is why I was hesitant to push it directly to your branch. I think that this is a much better solution in the long run, but if you don't have the time at the moment, you can of course also keep it in mind for later.

mattiarighi commented:
Since it requires a lot of changes to the NCL part, the earlier we do this the better.
Let me give it a try and see how much work it is (especially for other users who have already started their porting in other branches).

mattiarighi commented:
I have tried this with a simple recipe using 4 variable and 3 datasets.
What I get in the NCL interface is:
- an `input_file_info` list with 12 elements (4 times 3)
- a `dataset_info` list with 12 elements
- a `variable_info` list with 4 elements

It will need a lot of playing with the indexes in order to extract the correct information for each dataset and variable (for example in the `data_handling.ncl` routines). I do not see any simplification with this approach, or am I missing something?

bouweandela commented:
Looking at data_handling.ncl, I think you could get rid of all indices in there completely, I would write `read_data` like this:
```
ndef("read_data")
function read_data(info_item[1]:logical)
;
; Arguments
;    info_item: logical describing the current input file as defined in the
;    <variable>_info.ncl file in the preproc directory
;
; Return value
;    The actual data array with all metadata.
;
; Description
;    Retrieve the variable data and metadata based on the input_file_info
;    logical defined in the <variable>_info.ncl file automatically generated
;    by the preprocessor and stored in the preproc/<preproc_id>/ subdirectory.
;
; Caveats
;
; References
;
; Modification history
;   20180212_A-righ_ma: written.
;
local funcname, scriptname, infile, fin
begin

  funcname = "read_data"
  scriptname = "interface_scripts/data_handling.ncl"
  enter_msg(scriptname, funcname)

  infile = info_item@filename

  ; Check existance and supported format
  if (.not.isfilepresent(infile)) then
    error_msg("f", scriptname, funcname, "input file " + infile + \
              " not found or not a supported format")
  end if

  ; Open file
  log_info("Reading in file = " + infile)
  fin = addfile(infile, "r")

  ; Check that variable is in the file
  if (.not. isfilevar(fin, info_item@short_name)) then
    error_msg("f", scriptname, funcname, "variable " + variable + \
              " not found in file " + infile)
  end if

  ; Read data
  data = fin->$info_item@short_name$

  leave_msg(scriptname, funcname)
  return(data)

end

```

For conveniently selecting sublists to loop over (e.g. corresponding to a single variable or dataset) from the input_file_info list, you could implement functions similar to `select_metadata`, `group_metadata`, `sorted_metadata` and `sorted_group_metadata` in https://github.com/ESMValGroup/ESMValTool/blob/version2_development/esmvaltool/diag_scripts/shared/_base.py.

schlunma commented:
From a python user's point of view who is still new to NCL this looks very good. This way data extraction is much easier and clearer and also really similar to the python diagnostics.

I will test it as soon as you are ready. Maybe you could implement a small example diagnostic? Then it is easier to get used to the new syntax :smile:

mattiarighi commented:
Sure, the first thing to adjust will be the NCL example diagnostic.

mattiarighi commented:
> It would probably be useful to have a function group_metadata_by_attribute

We can implement it once there is a use case in one of the NCL diagnostics.
At the moment I don't think this feature is needed.

> I'm not too keen on the function select_metadata_by_index, because it encourages the use of numerical indices instead of more human readable variable names.

I will remove it. Note however that in the NCL diags, variables are identified by their index and not by their name. This is because they are written to be flexible and work with different input variables. In diagnostics with multiple variables, their order is also important (see [this example](https://github.com/ESMValGroup/ESMValTool/blob/Anav13JClim_v2/esmvaltool/diag_scripts/carbon_cycle/two_variables.ncl#L1) which @schlunma is implementing in #473).

mattiarighi commented:
@schlunma and @axel-lauer I have revised the NCL example diagnostic in `recipes/examples/recipe_ncl.yml`.

Please have a look and let me know if something is unclear.

The approach is now identical for single or multiple variable diagnostics, you just change the `variable_info` list index to access the variable you need.

It required (and will require) quite some change in the shared scripts, but it is also a good opportunity to get rid of unused code and clean them up.

Next step will be to adapt the perfmetrics.

Once this is merged we need to inform the users that are currently developing/porting NCL diagnostics and probably provide some support with these changes.

schlunma commented:
Looks good, I like the new approach, but as you said, this will require a lot of work. Can I already start to adapt my already ported diagnostics or is it better to wait until this PR is merged?

schlunma commented on esmvaltool/interface_scripts/interface.ncl:
```diff
@@ -34,16 +39,111 @@ begin
-    field_types = (/variable_info@field/)
-  end if
+  ; Define current script name as global variable (following Python syntax)
+  _DIAG_SCRIPT = diag_script_info@script
```
One little comment: in my understanding leading underscores in python mean that something is "private" and should not be used by the user. This is not the case for these variables, so I suggest removing the underscores.
```suggestion
  DIAG_SCRIPT = diag_script_info@script
```

mattiarighi replied:
Good point. Let's just use upper-case names for global variables.

bouweandela commented:
> We can implement it once there is a use case in one of the NCL diagnostics.
At the moment I don't think this feature is needed.

The diagnostic by @schlunma that you refer to has exactly this use case:

https://github.com/ESMValGroup/ESMValTool/blob/c2c6a361c21e642bdd185b0ae613cafce178c9aa/esmvaltool/diag_scripts/carbon_cycle/two_variables.ncl#L129



> Note however that in the NCL diags, variables are identified by their index and not by their name. This is because they are written to be flexible and work with different input variables.

With the function `select_metadata_by_name` you can just use the short name of the variable as index. There is no need to hard-code the variable short_name any more than there is to hard-code a numerical index, you can just store it in a variable. This makes debugging a lot easier, because if you print short_name you immediately know what variable you are dealing with, whereas with numerical indices you need an extra step to go from number to variable name.


> In diagnostics with multiple variables, their order is also important (see this example which @schlunma is implementing in #473).

In that case I recommend implementing a sorting function for `input_file_info` lists, e.g. called `sort_metadata_by_attribute`, similar to 

https://github.com/ESMValGroup/ESMValTool/blob/c3c61b5341037d01c776c3524c0dd4c767507a3d/esmvaltool/diag_scripts/shared/_base.py#L77

to ensure that `input_file_info` is always in the same order. At least with the current implementation, there is no guarantee on the order that the diagnostic script will receive variables in (this can be changed of course, but supplying a sorting function is probably safest and more flexible for users too if they want a different order than the default one.)

mattiarighi commented:
Sorry, but now I'm completely confused about this new implementation.
Consider a recipe with:

```
variables:
  var1:
    preprocessor: ...
    mip: ...
    field: ...
  var2:
    preprocessor: ...
    mip: ...
    field: ...
```

then in the NCL script I would expect `variable_info[0]@short_name = var1` and `variable_info[1]@short_name = var2`.

This seems to be the case (at least according to my tests, with 4 variables).

If not, we have to rethink it, because as I said, the NCL diagnostics with multiple variables often rely on their order, for example they would plot `var1` on the x-axis vs. `var2` on the y-axis.

bouweandela commented:
There is no fixed order of variables in `variable_info` in the current implementation and also not in the new implementation in this branch. This has never been a requirement, since the NCL interface only supported one variable anyway and Python diagnostics do not rely on this kind or ordering (a good example of why it is not a good idea to use numerical indices when not strictly necessary). If you need this feature, we need to implement it.

I already spotted one reason why if it may not work at the moment: the variables are stored in a map (Python dict) in the recipe. From the [Python documentation](https://docs.python.org/3.7/library/stdtypes.html#mapping-types-dict): "Changed in version 3.7: Dictionary order is guaranteed to be insertion order. This behavior was implementation detail of CPython from 3.6.". So that means that it will probably work in Python 3.6 and 3.7, but not in older versions of Python.

mattiarighi commented:
Using python 3.6.4 the order is preserved, that's why I was confused by your comment.

I think this feature is important in general, not only in NCL.

We've always been trying to write diagnostics than can be applied to various input variables.
For example, in #696 @ruthlorenz implemented a diagnostic where multi-model mean precipitation change (variable 1) is scaled by global temperature change (variable 2). This kind of diagnostic could be applied also to other variable pairs, but if the order is not fixed it makes no sense.

@axel-lauer do you have further examples which we could discuss here?

bouweandela commented:
In that case you could specify the names of the variables to use for change/scale in the diagnostic script settings section in the recipe. That seems a bit more robust than relying on the order.

mattiarighi commented:
This is also an option, but it would be error prone, in particular for NCL users with no experience with dictionaries, who would consider the `variable_info` list to be ordered and write their scripts accordingly. This is even more likely for v1 users.

Also consider that NCL lists are not so flexible, and most of the functionalities that make them profitable in python are not available in NCL. Of course we could add a set of convenience functions for that, but again is a matter of code usability for non experts.

mattiarighi commented:
NCL-based recipes to be adapted and tested with this improved interface:
- [ ] recipe_clouds_bias.yml
- [ ] recipe_clouds_ipcc.yml
- [ ] recipe_flato13ipcc.yml
- [ ] recipe_lauer13jclim.yml
- [ ] recipe_perfmetrics_CMIP5.yml

@axel-lauer kindly agreed to work on the first four, I will do the last one.

axel-lauer commented:
I have tried this and it seems that NCL 6.4.0 cannot handle lists of lists correctly resulting in error such as segmentation violation and/or other problems. It is, however, not entirely clear to me why we need lists of lists. NCL is not Python, so wouldn't a list of "normal" variables (that can be arrays) plus atttributes do?

mattiarighi commented:
Specifically for `select_metadata_by_atts`, lists are needed to pass attribute=value pairs, since values can be of any type and NCL arrays do not allow mixed types. For example, "short_name" = "ta" (string) and "start_year" = 2005 (integer).

More generally, the idea of using NCL lists instead of arrays to store the metadata was proposed by @bouweandela to have a consistent interface for all supported language (NCL, python and R).

But if this does not work, or if it turns out to be too complex from a NCL-user's perspective, we should reopen #682 and use the previous approach.

bouweandela commented:
Hi Mattia,

in this case, I think you can just use a logical instead of a list of lists for `atts`, e.g.:
```ncl
atts = True
atts@short_name = "tas"
atts@dataset = "ERA-Interim"
atts@mip = "Amon"
```
It's slightly less nice than a list of lists, but at least it should work.

bouweandela commented:
You're right about length, but I like the symmetry: the list of logicals with attributes you get back are in the same format. Maybe it's just the best you can do with NCL. If you really want shorter, you could instead write the function
```ncl
select_metadata_by_attribute(input_file_info, key, value)
```
because in most cases you will probably only use it with one attribute anyway, and if you need more, you can always chain calls like so (assuming NCL doesn't fail on that too):
```ncl
info = select_metadata_by_attribute(input_file_info, "short_name". var0)
info = select_metadata_by_attribute(info, "field", field0)
info = select_metadata_by_attribute(info, "dataset", reference_dataset0)
```

mattiarighi commented:
@axel-lauer I was thinking that the function `listatts2array` should be moved to `interface.ncl` together with the other convenience function. This would be more consistent and would also ensure that it is always loaded in the diagnostics. We should also rename it to `metadata_att_as_list`.

mattiarighi commented:
@bouweandela there is another issue with the new interface. Some `diag_script_info` and `input_file_info` attributes which were previously converted to arrays are now passed as lists. For example, the contour levels:
```
diag_script_info@abs_levs = [/200, 210, 220, 230, 240, 250, 260, 270, 280, 290, 300/]
```
or the input file:
```
diag_script_info@input_files = [/"/work/bd0080/b309057/ESMValTool/v2_output/recipe_perfmetrics_CMIP5_20181120_084621/preproc/taZONAL_ppALL_ta/ta_info.ncl"/]
``` 
In NCL, this is confusing, since the list items are not variables in this case, but actual values.

Can we change it?

mattiarighi commented on esmvaltool/_task.py:
```diff
@@ -112,51 +112,68 @@ def _log_resource_usage():
         else:
-            txt = str(value)
-        return txt
+            txt += '[/{}/]'.format(', '.join(_py2ncl(v) for v in value))
```
```suggestion
            txt += '(/{}/)'.format(', '.join(_py2ncl(v) for v in value))
```

mattiarighi replied:
This should solve the problem.

bouweandela commented:
> In NCL, this is confusing, since the list items are not variables in this case, but actual values.

> Can we change it?

No, that will lead to unexpected problems because arrays cannot hold values of different types, while lists can. So this will fail in some cases, e.g. when a user defines a list with items of different types in the recipe that is then converted to an invalid NCL array. If you need to convert some lists to arrays, you should do so where it is needed in the NCL code.

mattiarighi commented:
> They are variables with a value assigned to them. Other variables (i.e. not stored in a list) in your NCL scripts also just represent the value that is stored in them, so why is this confusing?

Because, as I said already several times, NCL is a nightmare with lists.

For example, I can no longer pass a setting such as the contour levels to the corresponding plot resources without converting it to array before. 
The same happens when writing NetCDF files in the diagnostics, since the list type is not supported as a variable attribute in NetCDF.

> If you need to convert some lists to arrays, you should do so where it is needed in the NCL code.

This is going to make NCL very user unfriendly, I'm stronly against this.

schlunma commented:
> > In NCL, this is confusing, since the list items are not variables in this case, but actual values. Can we change it?
> 
> No, that will lead to unexpected problems because arrays cannot hold values of different types, while lists can. So this will fail in some cases, e.g. when a user defines a list with items of different types in the recipe that is then converted to an invalid NCL array. If you need to convert some lists to arrays, you should do so where it is needed in the NCL code.

In my opinion, we should definitely use NCL arrays here:

1. I do not know of any use case where we need settings with different variable types for NCL diagnostics.

2. NCL lists are not that flexible as python lists, as @mattiarighi pointed out several times. Additionally, the implementation is not very robust which may lead to unexpected problems and bugs in the future.

3. We could include a check in python if the settings include different variable types. If this is the case, print an `NCL does not support this ...` error message.

This way, we can keep the diagnostics simple (no need to convert list -> array every time settings are used) and hopefully do not run into further problems with NCL lists.

axel-lauer commented:
I completely second @schlunma. I think using arrays instead of lists in NCL is the *only* feasible way at this moment as (1) we are not able to recode all existing NCL diagnostics and (2) using lists makes things too complicated for typical scientific NCL users - even when providing some convenience functions. I think we simply have to accept that NCL is not Python and thus cannot support all Python features.

In addition, I would argue that we need *ordered* dictionaries so that the order of variables in the recipes is preserved when passing the info to NCL. This is needed for NCL diagnostics, which need to know which of the variables is the "main" variable and which are "additional" variables. Introducing additional configuration variables in the recipe is in my opinion too error prone as technically redundant information would need to be provided. If this means dropping Python 2 support, fine with me.

**Both of the issues need to be settled very soon as the work needed to adapt all NCL code already starts to spiral out of control!**

bjoernbroetz commented:
>... in particular for NCL users with no experience with dictionaries, who would consider the variable_info list to be ordered and write their scripts accordingly ...

For the problem of the ordered variables, I would like to propose that we add an optional attribute like `ncl_order` for the variables in the recipes that is then used when `variable_info` is defined. Currently, some of the users implicitly assume that there is an order in the variable definition in the recipe, that is not there. But this might be something for a separate PR.

With respect to the list vs array problem: I can confirm, as @schlunma mentioned, the use-case "multiple data types grouped to one attribute" is currently not needed. Given the issues that our NCL-people experienced with the List type of NCL I agree that we should opt to arrays when we need ordered items in the NCL world.

bouweandela commented:
Regarding the list vs array issue: let's go for arrays then. I've just added a check that all elements are the same to avoid NCL syntax errors in case someone specifies mixed elements. Also, I've changed to code so an empty list is converted to a `_Missing` value, because it is not possible to create an empty array.

Regarding the variable ordering issue, I have created an actual issue for it  #724. However, I still think that it would be better to explicitly name what variable you use for what in the recipe. We also have `reference_dataset` to name the reference dataset, we do not assume it is the first dataset in the list or something like that.

mattiarighi commented:
If there are no further suggestions this is now ready to go. :racing_car:  :checkered_flag:
